torch/optim/_muon.py

```diff
@@ -0,0 +1,336 @@
+            momentum = group["momentum"]
+            beta1, beta2 = group["adamw_betas"]
+            eps = group["adamw_eps"]
+
+            muon_params: list[Tensor] = []
+            muon_grads: list[Tensor] = []
+            muon_momentum_bufs: list[Tensor] = []
+            other_params: list[Tensor] = []
+            other_grads: list[Tensor] = []
+            exp_avgs: list[Tensor] = []
+            exp_avg_sqs: list[Tensor] = []
+            steps: list[Tensor] = []
+
+            self._init_group(
+                group,
+                muon_params,
+                muon_grads,
+                muon_momentum_bufs,
+                other_params,
+                other_grads,
+                exp_avgs,
+                exp_avg_sqs,
+                steps,
+            )
+
+            _single_tensor_muon(
+                muon_params,
+                muon_grads,
+                muon_momentum_bufs,
+                other_params,
+                other_grads,
+                exp_avgs,
+                exp_avg_sqs,
+                steps,
+                lr=lr,
+                wd=wd,
+                momentum=momentum,
+                nesterov=group["nesterov"],
+                beta1=beta1,
+                beta2=beta2,
+                eps=eps,
+                msign_fn=self._msign_fn,
+                msign_fn_config=self._msign_fn_config,
+                adjust_lr_fn=self._adjust_lr_fn,
+            )
+        return loss
+
+
+def _single_tensor_muon(
+    muon_params: list[Tensor],
+    muon_grads: list[Tensor],
+    muon_momentum_bufs: list[Tensor],
+    other_params: list[Tensor],
+    other_grads: list[Tensor],
+    exp_avgs: list[Tensor],
+    exp_avg_sqs: list[Tensor],
+    steps: list[Tensor],
+    *,
+    lr: float,
+    wd: float,
+    momentum: float,
+    nesterov: bool,
+    beta1: float,
+    beta2: float,
+    eps: float,
+    msign_fn: Callable[[Tensor, BaseMsignFnConfig], Tensor],
+    msign_fn_config: BaseMsignFnConfig,
+    adjust_lr_fn: Callable[[float, torch.Size], float],
+) -> None:
+    lr = _to_scalar(lr)
+    for i, param in enumerate(muon_params):
+        grad = muon_grads[i]
+        assert grad.ndim == 2, "Param gradient must be a 2D matrix"
+
+        buf = muon_momentum_bufs[i]
+        buf.mul_(momentum).add_(grad)
+        if nesterov:
+            grad = grad.add(buf, alpha=momentum)
+        else:
+            grad = buf
+
+        update = msign_fn(grad, msign_fn_config)
+        adjusted_lr = adjust_lr_fn(lr, param.shape)
+        param.mul_(1 - lr * wd)
+        param.add_(update, alpha=-adjusted_lr)
+
+    for i, param in enumerate(other_params):
+        grad = other_grads[i]
+        exp_avg = exp_avgs[i]
+        exp_avg_sq = exp_avg_sqs[i]
+        step = steps[i]
+        step += 1
+
+        exp_avg.lerp_(grad, 1 - beta1)
+        exp_avg_sq.lerp_(grad.square(), 1 - beta2)
+
+        bias_correction1 = 1 - beta1 ** step.item()
+        bias_correction2 = 1 - beta2 ** step.item()
```

Comment:
.item() calls are expensive--why not:
```suggestion
        bias_correction1 = 1 - beta1 ** step
        bias_correction2 = 1 - beta2 ** step
```